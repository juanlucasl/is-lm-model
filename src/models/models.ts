/* SPDX-License-Identifier: Apache-2.0 */
export type FunctionFormula = (x: number) => number;

export type Variable = {
  value: number;
  setValue: (value: number) => void;
  label: string;
  increaseDescription?: string | JSX.Element;
  decreaseDescription?: string | JSX.Element;
};
