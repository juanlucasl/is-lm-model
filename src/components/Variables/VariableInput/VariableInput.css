/* SPDX-License-Identifier: Apache-2.0 */
.variable-input {
  align-items: center;
  display: grid;
  grid-template-columns: 5fr 1fr;
  grid-column-gap: 1rem;
}

.variable-input span {
  grid-row: 1 / span 2;
  justify-content: center;
}

.arrow {
  cursor: pointer;
  font-weight: bolder;
  font-size: 1.6rem;
  user-select: none;
}

.arrow__upwards {
  color: green;
}

.arrow__upwards:active {
  transform: translateY(-3px);
}

.arrow__downwards {
  color: red;
}

.arrow__downwards:active {
  transform: translateY(3px);
}
