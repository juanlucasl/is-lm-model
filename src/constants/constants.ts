/* SPDX-License-Identifier: Apache-2.0 */

export const GRAPH_SIZE = 500;

export const MAX_VALUE = GRAPH_SIZE;
export const MIN_VALUE = -100;

/**
 * Propension Marginal a Consumir: Indica como una unidad más de renta disponible afecta al
 * consumo.
 */
export const PROPENSION_MARGINAL_A_CONSUMIR = 0.8;

export const SENSIBILIDAD_DE_LA_DEMANDA_MONETARIA_AL_INTERES = 0.5;

/**
 * Lo que consumirían los individuos si su renta disponible fuese igual a cero.
 */
export const CONSUMO_AUTONOMO = 100;
export const DEMANDA_BASE_DE_DINERO = CONSUMO_AUTONOMO;
