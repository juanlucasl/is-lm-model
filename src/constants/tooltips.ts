const CONSUMPTION_UP = "&uarr; Consumo (C)";
const CONSUMPTION_DOWN = "&darr; Consumo (C)";

const INTEREST_RATE_UP = "&uarr; Tipo de Interés (i)";
const INTEREST_RATE_DOWN = "&darr; Tipo de Interés (i)";

const OUTPUT_UP = "&uarr; Producción o PIB (Y) (En equilibrio Y = Z (Demanda de bienes))";
const OUTPUT_DOWN = "&darr; Producción o PIB (Y) (En equilibrio Y = Z (Demanda de bienes))";

const IS_RIGHT = "&rarr; La Curva IS se desplaza hacia la derecha, reflejando un aumento en la Demanda Agregada";
const IS_LEFT = "&larr; La Curva IS se desplaza hacia la izquierda, reflejando un descenso en la Demanda Agregada";

export const TAXES_INCREASE_DESCRIPTION = `
  <p>
    El Estado cobra más impuestos y la gente tiene menor excedente de renta para demandar
    bienes
    <ul>
      <li>${CONSUMPTION_DOWN}</li>
      <li>${INTEREST_RATE_DOWN}</li>
      <li>${OUTPUT_DOWN}</li>
      <li>${IS_LEFT}</li>
    </ul>
  </p>
`;

export const TAXES_DECREASE_DESCRIPTION = `
  <p>
    El Estado cobra menos impuestos y la gente tiene mayor excedente de renta para demandar
    bienes
    <ul>
      <li>${CONSUMPTION_UP}</li>
      <li>${INTEREST_RATE_UP}</li>
      <li>${OUTPUT_UP}</li>
      <li>${IS_RIGHT}</li>
    </ul>
  </p>
`;

export const INVESTMENT_INCREASE_DESCRIPTION = `
  <p>
    Mayor demanda de bienes de capital, por lo tanto, es mayor la Demanda Agregada
    <ul>
      <li>${CONSUMPTION_UP}</li>
      <li>${INTEREST_RATE_UP}</li>
      <li>${OUTPUT_UP}</li>
      <li>${IS_RIGHT}</li>
    </ul>
  </p>
`;

export const INVESTMENT_DECREASE_DESCRIPTION = `
  <p>
    Menor demanda de bienes de capital, por lo tanto, es menor la Demanda Agregada
    <ul>
      <li>${CONSUMPTION_DOWN}</li>
      <li>${INTEREST_RATE_DOWN}</li>
      <li>${OUTPUT_DOWN}</li>
      <li>${IS_LEFT}</li>
    </ul>
  </p>
`;

export const PUBLIC_EXPENDITURE_INCREASE_DESCRIPTION = `
  <p>
    El Estado demanda más bienes y servicios, por lo tanto, es mayor la Demanda Agregada
    <ul>
      <li>${CONSUMPTION_UP}</li>
      <li>${INTEREST_RATE_UP}</li>
      <li>${OUTPUT_UP}</li>
      <li>${IS_RIGHT}</li>
    </ul>
  </p>
`;

export const PUBLIC_EXPENDITURE_DECREASE_DESCRIPTION = `
  <p>
    El Estado demanda menos bienes y servicios, por lo tanto, es menor la Demanda Agregada
    <ul>
      <li>${CONSUMPTION_DOWN}</li>
      <li>${INTEREST_RATE_DOWN}</li>
      <li>${OUTPUT_DOWN}</li>
      <li>${IS_LEFT}</li>
    </ul>
  </p>
`;

export const MONEY_SUPPLY_INCREASE_DESCRIPTION = `
  <p>
    Mayor oferta de dinero hace bajar su precio, reflejado en el interés
    <ul>
      <li>${INTEREST_RATE_DOWN}</li>
    </ul>
  </p>
`;

export const MONEY_SUPPLY_DECREASE_DESCRIPTION = `
  <p>
    Menor oferta de dinero hace subir su precio, reflejado en el interés{" "}
    <ul>
      <li>${INTEREST_RATE_UP}</li>
    </ul>
  </p>
`;
