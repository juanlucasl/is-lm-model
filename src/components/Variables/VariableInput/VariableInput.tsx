/* SPDX-License-Identifier: Apache-2.0 */
import "./VariableInput.css";
import React from "react";
import { Tooltip } from "react-tooltip";
import { Variable } from "../../../models";

interface InputProps {
  variable: Variable;
}

const STEP = 5;

const removeWhitespaceSpecialCharsAndMakeLower = (string: string) =>
  string.replace(/[\W\s]/gi, "").toLowerCase();

const VariableInput: React.FC<InputProps> = ({
  variable: { value, setValue, label, increaseDescription, decreaseDescription },
}) => (
  <div className="variable-input">
    <span>{label}</span>
    <div
      className="arrow arrow__upwards"
      data-tooltip-html={increaseDescription as string}
      id={`${removeWhitespaceSpecialCharsAndMakeLower(label)}-increase`}
      onClick={() => setValue(value + STEP)}
    >
      &uarr;
    </div>
    <div
      className="arrow arrow__downwards"
      data-tooltip-html={decreaseDescription as string}
      id={`${removeWhitespaceSpecialCharsAndMakeLower(label)}-decrease`}
      onClick={() => setValue(value - STEP)}
    >
      &darr;
    </div>
    <Tooltip
      anchorId={`${removeWhitespaceSpecialCharsAndMakeLower(label)}-increase`}
      className="variable-input__tooltip"
      place="right"
    />
    <Tooltip
      anchorId={`${removeWhitespaceSpecialCharsAndMakeLower(label)}-decrease`}
      className="variable-input__tooltip"
      place="right"
    />
  </div>
);

export default VariableInput;
