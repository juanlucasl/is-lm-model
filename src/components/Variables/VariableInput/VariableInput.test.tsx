/* SPDX-License-Identifier: Apache-2.0 */
import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import VariableInput, { STEP } from "./VariableInput";
import { Variable } from "../../../models";
import { MAX_VALUE, MIN_VALUE } from "../../../constants";

const mockVariable: Variable = {
  value: 0,
  setValue: jest.fn(),
  label: "Mock Variable Input",
  increaseDescription: "Mock increase description",
  decreaseDescription: "Mock decrease description",
};

describe("VariableInput", () => {
  it("renders correctly", () => {
    const { container } = render(<VariableInput variable={mockVariable} />);
    expect(container).toMatchSnapshot();
  });

  it("should render a label and two arrows", () => {
    render(<VariableInput variable={mockVariable} />);
    expect(screen.getByText(mockVariable.label)).toBeVisible();
    expect(screen.getByRole("button", { name: /\+/i })).toBeVisible();
    expect(screen.getByRole("button", { name: /-/i })).toBeVisible();
  });

  it("should call the setValue function with the new value when the increase arrow is clicked", () => {
    render(<VariableInput variable={mockVariable} />);
    userEvent.click(screen.getByRole("button", { name: /\+/i }));
    expect(mockVariable.setValue).toHaveBeenCalledWith(STEP);
  });

  it("should call the setValue function with the new value when the decrease arrow is clicked", () => {
    render(<VariableInput variable={mockVariable} />);
    userEvent.click(screen.getByRole("button", { name: /-/i }));
    expect(mockVariable.setValue).toHaveBeenCalledWith(-STEP);
  });

  it("should disable the increase arrow when the value is at the maximum", () => {
    render(<VariableInput variable={{ ...mockVariable, value: MAX_VALUE }} />);
    const arrowIncrease = screen.getByRole("button", { name: /\+/i });

    userEvent.click(arrowIncrease);
    expect(mockVariable.setValue).not.toHaveBeenCalled();
    expect(arrowIncrease).toHaveAttribute("aria-disabled", "true");
    expect(arrowIncrease).toHaveClass("arrow--disabled");
  });

  it("should disable the decrease arrow when the value is at the minimum", () => {
    render(<VariableInput variable={{ ...mockVariable, value: MIN_VALUE }} />);
    const arrowDecrease = screen.getByRole("button", { name: /-/i });

    userEvent.click(arrowDecrease);
    expect(mockVariable.setValue).not.toHaveBeenCalled();
    expect(arrowDecrease).toHaveAttribute("aria-disabled", "true");
    expect(arrowDecrease).toHaveClass("arrow--disabled");
  });
});
