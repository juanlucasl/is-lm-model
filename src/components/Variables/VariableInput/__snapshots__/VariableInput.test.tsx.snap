// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VariableInput renders correctly 1`] = `
<div>
  <div
    class="variable-input"
  >
    <span>
      Mock Variable Input
    </span>
    <div
      aria-disabled="false"
      aria-label="+ Mock Variable Input"
      class="arrow arrow__upwards "
      data-testid="arrow-increase"
      data-tooltip-html="Mock increase description"
      id="mockvariableinput-increase"
      role="button"
    >
      ↑
    </div>
    <div
      aria-disabled="false"
      aria-label="- Mock Variable Input"
      class="arrow arrow__downwards "
      data-testid="arrow-decrease"
      data-tooltip-html="Mock decrease description"
      id="mockvariableinput-decrease"
      role="button"
    >
      ↓
    </div>
  </div>
</div>
`;
